ADR/0033-provisioning-test-resources.md

```diff
@@ -0,0 +1,277 @@
+
+The problem of provisioning test resources/environments can be broken down into a few questions:
+
+1. How can resources (OpenShift clusters, to start) be provisioned, efficiently, without exposing
+   shared cloud/infra credentials to the end user?
+1. In case the user requires more customization than what's provided by shared configuration, how
+   can they provide their own (including infra credentials)?
+1. How does the end user request resources from an integration `Pipeline`?
+
+Provisioning of infrastructure can consume some significant compute resources itself. All possible
+solutions must also account for the challenge of scalability in production.
+
+### Cluster Provisioning
+
+[Hive] may be used to provision OpenShift clusters. It's widely used in OpenShift CI testing,
+supports hibernated cluster pools for quick (5-7m), cost-efficient, allocation and is maintained
+and distributed by Red Hat on OperatorHub. Architecture support for provisioned clusters is
+determined by the infra provider for which Hive supports all the most popular options. The
+[scaling characteristics][HiveScaling] of Hive on a single cluster are well documented with some
+known upper limits (max of ~1000 provisioned clusters per management cluster).
+
+[Hypershift] allows for creating control planes at scale with reduced cost and provisioning time
+when compared to Hive. But unlike Hive, the version of the management cluster dictates the
+cluster versions which can be provisioned. For example, Hypershift deployed on top of OpenShift 4.14
+only allows for requesting control planes for version 4.12-4.14. Hypershift scales better than Hive
+though since the control plane is deployed as pods on worker nodes. It's currently available as
+part of OpenShift TechPreview and supports deploying 64-bit x86 and 64-bit ARM `NodePools`. 
+
+[Cluster API][CAPI] is an intriguing option for provisioning Kubernetes and OpenShift clusters but
+expected to remain in OpenShift TechPreview throughout 2024. A dedicated
+management cluster separate from application workloads is recommended when deploying this in
+production.
+
+The Cluster as a Service ([CaaS]) Operator provides self-service cluster provisioning using
+additional guardrails like custom templates and quotas. CaaS supports [Hive] and [Hypershift] for
+the cluster creation process. It uses Helm as a templating engine which makes it quite flexible.
+This Operator also provides the option to apply the resources, generated from a template, to the
+namespace alongside the related `ClusterTemplateInstance` or to a common namespace which is
+necessary when protecting shared credentials.
+
+## Decision
+
+### Cluster Provisioning
+
+We will use the ([CaaS]) Operator to orchestrate the process of provisioning OpenShift clusters.
+Users will create `ClusterTemplateInstances` referencing `ClusterTemplates` which will be
+maintained by Konflux admins. By default, the templates will reference infra/pull/ssh
```

Comment:
What will be the design considerations behind the maintenance of the `ClusterTemplates` by admins? Is there any sense of the knobs that might be available for users?

Reply:
ClusterTemplates are expected to be managed in a git repo and configured via infra-deployments to keep maintenance overhead low. We'll obviously need to make adjustments to these templates as requested or required although it might be a good opportunity to lean on community support for enhancements.

The initial set of parameters for each template may be minimal at first (arch and version are the obvious one to me) and I expect early adopters will help determine further requirements.